sanitizers/src/main/java/com/code_intelligence/jazzer/sanitizers/XpathInjection.kt

```diff
@@ -0,0 +1,92 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.code_intelligence.jazzer.sanitizers
+
+import com.code_intelligence.jazzer.api.FuzzerSecurityIssueHigh
+import com.code_intelligence.jazzer.api.HookType
+import com.code_intelligence.jazzer.api.Jazzer
+import com.code_intelligence.jazzer.api.MethodHook
+import com.code_intelligence.jazzer.api.MethodHooks
+import java.lang.invoke.MethodHandle
+
+/**
+ * Detects Xpath injections.
+ *
+ * Untrusted input has to be escaped in such a way that queries remain valid otherwise an injection
+ * could be possible. This sanitizer guides the fuzzer to inject insecure characters. If an exception
+ * is raised during execution the fuzzer was able to inject an invalid pattern, otherwise all input
+ * was escaped correctly.
+
+ * False positives can be possible but should bugs in the code just not an injection. e.g.
+ * "If expression contains any XPathFunctions, they must be available via the XPathFunctionResolver. An XPathExpressionException will be thrown if the XPathFunction cannot be resovled with the XPathFunctionResolver."
+ * "The parameter item represents the context the XPath expression will be operated on. The type of the context is implementation-dependent. If the value is null, the operation must have no dependency on the context, otherwise an XPathExpressionException will be thrown."
+ * https://docs.oracle.com/en/java/javase/17/docs/api/java.xml/javax/xml/xpath/XPath.html
+ *
+ * Checking the number of quotation marks to be uneven should indicate injection instead of a false positive
+ */
+@Suppress("unused_parameter", "unused")
+object XpathInjection {
+
+    // Characters that should be escaped in user input.
+    // https://owasp.org/www-community/attacks/XPATH_Injection
+    private const val CHARACTERS_TO_ESCAPE = "'\""
+
+    private val XPATH_SYNTAX_ERROR_EXCEPTIONS = listOf(
+        "javax.xml.xpath.XPathExpressionException",
+    )
+
+    @MethodHooks(
+        MethodHook(type = HookType.REPLACE, targetClassName = "javax.xml.xpath.XPath", targetMethod = "compile"),
+        MethodHook(type = HookType.REPLACE, targetClassName = "javax.xml.xpath.XPath", targetMethod = "evaluate"),
+        MethodHook(type = HookType.REPLACE, targetClassName = "javax.xml.xpath.XPath", targetMethod = "evaluateExpression")
+    )
+
+    @JvmStatic
+    fun checkXpathExecute(method: MethodHandle, thisObject: Any?, arguments: Array<Any>, hookId: Int): Any {
+
+        if (arguments.isNotEmpty() && arguments[0] is String) {
+            val query = arguments[0] as String
+            Jazzer.guideTowardsContainment(query, CHARACTERS_TO_ESCAPE, hookId)
+        }
+        return try {
+            method.invokeWithArguments(thisObject, *arguments)
+        } catch (throwable: Throwable) {
+            if (XPATH_SYNTAX_ERROR_EXCEPTIONS.contains(throwable.javaClass.name) && !isEvenNumberOfQuotationMarks(arguments[0] as String)) {
```

Comment:
I delved into the `java.xml` source code to find a better signal that a given `XPathExpressionException` actually arises from a syntax error.

The actual parsing happens in https://github.com/openjdk/jdk/blob/master/src/java.xml/share/classes/com/sun/org/apache/xpath/internal/compiler/XPathParser.java, which seems to always call the `error` function in case of a parsing error. This function creates a `TransformerException` at https://github.com/openjdk/jdk/blob/master/src/java.xml/share/classes/com/sun/org/apache/xpath/internal/compiler/XPathParser.java#L644, which code higher up the call stack wraps into an `XPathExpressionException`, together with essentially any other kind of exception (for example at https://github.com/openjdk/jdk/blob/e5edb10dc56d9edac8e050e0f8e6c116743975d6/src/java.xml/share/classes/com/sun/org/apache/xpath/internal/jaxp/XPathImpl.java#L174).

I briefly looked through the other occurrences of `new TransformerException` and it seems that we could just catch `XPathExpressionException` instead of `Throwable` and report a finding if the innermost cause of that exception is a `TransformerException`.

Reply:
While I did not take a closer look at this approach yet, the reason I had before to not dive deeper into OpenJDKs code was that I am not sure how standardized differend implimentation of the JDK are and if we could trust them to internally be/stay the same. If that is not a problem this approach seems to be much nicer than my solution.

Reply:
It looks like the OpenJDK implementation is just a vendored version of an Apache library, so I would assume other JDKs use the same one. While Jazzer itself tries to stay out of JDK internals, hook implementations can't realistically do this (see our regex hooks, which make extensive assumptions about the implementation). 

If something works well with OpenJDK I would say go with that and revisit once it leads to problems with other JDKs. In this particular case the exception type we would be testing for is part of the public API surface, so there is a good chance this will even work across JDKs.

Reply:
Ok so after playing around with it, your proposed approach seems to work fine. The only problem I see right now is that the error() func which than creates the TransformerException is called from several places that don't necessary imply an injection. So there still could be some false positives. 

From my point of view the isEvenNumberOfQuotationMarks() approach before will prevent all false positives but not detect an injection if quotation marks in input are sanitized and the programm allows the input to directly build a querry.
The new approach probably detects all injections but has some false positives. I can't really assess how probable these are in production code.

What is your opinion ?

Reply:
I also tried to figure out where exactly `TransformerException` are thrown, and as far as I can see they are always related to parsing, at least in `XPathParser`. Other classes throw `TransformException`s as well, though.

IMHO it's ok to assume that queries are generally valid and would pass internal checks. In that case we could check the exception type as @fmeum suggested. 

Which one would you prefer?

Reply:
I would vote for the TransformerException approach then. It seems like a nicer solution. If there are some unexpected false positives, we could still fall back to the other approach.